File: backend/app/schemas/batch_upload.py
from __future__ import annotations

from datetime import datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel

from app.core.validators import CamelCaseModel
from app.schemas.user import UserResponse


class BatchUploadCreate(BaseModel):
    entity_type: str = "project"
    entity_id: Optional[UUID] = None
    category: Optional[str] = None
    building: Optional[str] = None
    floor: Optional[str] = None


class BatchUploadFileResponse(CamelCaseModel):
    id: UUID
    filename: str
    file_type: Optional[str] = None
    file_size: Optional[int] = None
    storage_path: str
    uploaded_at: datetime


class BatchUploadResponse(CamelCaseModel):
    id: UUID
    project_id: UUID
    total_files: int
    processed_files: int
    failed_files: int
    status: str
    metadata_json: Optional[dict] = None
    created_at: datetime
    completed_at: Optional[datetime] = None
    uploader: Optional[UserResponse] = None


class BatchUploadStatusResponse(CamelCaseModel):
    id: UUID
    project_id: UUID
    total_files: int
    processed_files: int
    failed_files: int
    status: str
    metadata_json: Optional[dict] = None
    created_at: datetime
    completed_at: Optional[datetime] = None
    uploader: Optional[UserResponse] = None
    files: list[BatchUploadFileResponse] = []

File: frontend/src/api/batchUpload.ts
import { apiClient } from './client'

export interface BatchUploadResponse {
  id: string
  projectId: string
  totalFiles: number
  processedFiles: number
  failedFiles: number
  status: 'pending' | 'processing' | 'completed' | 'failed'
  metadataJson: Record<string, string> | null
  createdAt: string
  completedAt: string | null
  uploader: {
    id: string
    full_name: string
  } | null
}

export interface BatchUploadFileResponse {
  id: string
  filename: string
  fileType: string | null
  fileSize: number | null
  storagePath: string
  uploadedAt: string
}

export interface BatchUploadStatusResponse extends BatchUploadResponse {
  files: BatchUploadFileResponse[]
}

export interface BatchUploadMetadata {
  entityType?: string
  entityId?: string
  category?: string
  building?: string
  floor?: string
}

export async function uploadBatch(
  projectId: string,
  files: File[],
  metadata?: BatchUploadMetadata,
): Promise<BatchUploadResponse> {
  const formData = new FormData()
  files.forEach((file) => formData.append('files', file))

  if (metadata?.entityType) formData.append('entity_type', metadata.entityType)
  if (metadata?.entityId) formData.append('entity_id', metadata.entityId)
  if (metadata?.category) formData.append('category', metadata.category)
  if (metadata?.building) formData.append('building', metadata.building)
  if (metadata?.floor) formData.append('floor', metadata.floor)

  const response = await apiClient.post<BatchUploadResponse>(
    `/projects/${projectId}/batch-uploads`,
    formData,
    { headers: { 'Content-Type': 'multipart/form-data' } },
  )
  return response.data
}

export async function getBatchStatus(
  projectId: string,
  batchId: string,
): Promise<BatchUploadStatusResponse> {
  const response = await apiClient.get<BatchUploadStatusResponse>(
    `/projects/${projectId}/batch-uploads/${batchId}`,
  )
  return response.data
}

export async function listBatches(
  projectId: string,
): Promise<BatchUploadResponse[]> {
  const response = await apiClient.get<BatchUploadResponse[]>(
    `/projects/${projectId}/batch-uploads`,
  )
  return response.data
}
